### ISSUE TITLE
Reset global phase of the full state simulator

### ISSUE THREAD

**Issue by jgukelberger** [microsoft/qsharp-runtime#35]
`ResetAll` on all qubits and/or deallocating all qubits should reset the global phase of the full state simulator. 

The global phase of the wavefunction is arbitrary and has no effect on measurements. However, it is reflected in the `DumpMachine` and `DumpRegister` outputs. Currently, the simulator retains a nontrivial phase even when the user goes through a sequence that should reset it to a clean state, such as deallocating all qubits. This leads to the confusing behavior that the same gate sequence applied to supposedly identical initial states (created, e.g., by releasing and allocating all qubits or applying `ResetAll` to all qubits) leads to different machine states, as observed through the dumps. It also makes the process of checking dumps against reference states for debugging purposes more involved than needed.

**Comment by anpaz-msft** [comment]
I have a bit of concerned about doing this for performance reasons, but might be ok. 
If we're doing this during `Reset` we should probably expose implemented as an Intrinsic for the simulator. An alternative is to implement this during measurement (`M`). After all, after measurement the global phase should also be reset `0`, correct?

Regarding your last comment, I believe that what you're saying is that if we do this during `Dump`, we'll be loosing some of the unitary information, correct?

This looks like a good enhancement, I'm going to mark it as "looking for help" as we don't have resources right now to work on it.

**Comment by jgukelberger** [comment]
Thanks for the comment! 

I'm not sure I understand your performance concern: You would only ever reset the phase when the wavefunction is a purely classical computational basis state, so you'd have to set a single amplitude to 1.0. I don't see how this would be significant overhead compared to qubit measurement or allocation.

Regarding phase normalization during `Dump`: Yes, this would essentially introduce an arbitrary, state-dependent phase shift in every `Dump` operation, which may again create confusion when users compare different dumps. Imagine a user applying the same operation `U` to different states and examining the register dumps after each of U|0>, U|1>, U(|0>+|1>)/sqrt(2). From the dumps, they'd conclude that U|0> + U|1> != U(|0>+|1>), i.e. that additivity is violated.

**Comment by crazy4pi314** [comment]
I think the problem here is the assumption that people _should_ be looking at the `Dump` to understand things. The problem with that is that it trains people that they will always have that resource as a debugging tool, when in reality, you cannot look at the `Dump` of an actual qubit. IMHO I think its a good thing that the dumps would not match in the example that @jgukelberger gave above because it should be a teaching moment to show that there is information in them that you cannot directly observe.

**Comment by jgukelberger** [comment]
@crazy4pi314 - I think I understand where you're coming from, and there may be a valid point in saying the `Dump` operations should not be used in a pedagogic introduction to quantum computing; but applying that argument it would seem like none of the quantum register dumps should be available to the user. However, we do have them because we'd like to provide developers the best debugging tools possible.
My argument is that inconsistent handling of the global phase makes this particular debugging tool less useful than it could be. After all, a big advantage of developing on a simulated/emulated system (in general, independent of the quantum context) is the access to additional control and state information that the real system doesn't expose.

**Comment by anpaz-msft** [comment]
I agree with @jgukelberger, the whole purpose of having a simulator is to allow people to try and debug their algorithms; `Dump` is a debugging tool and keeping this global phase makes debugging harder. 
I also agree with @crazy4pi314 in that people should not rely on it to understand what the algorithm is doing, but as @jgukelberger said, `Dump` should be used as a debugging and not as a pedagogic tool. I do think we should change it.
